src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java

```diff
@@ -252,6 +252,23 @@ private void importFromGcs(String projectId, TableReference tableRef, @Nullable
           .setProjectId(tableRef.getProjectId())
           .setTableId(temporaryTableName);
         loadConfig.setDestinationTable(temporaryTableReference);
+      } else if (Operation.UPSERT.equals(operation)) {
```

Comment:
1. It seems like we should use condition "!Operation.INSERT.equals(operation)". I have take a look the definition for "update". 

Update - records that match on Table Key will be updated in the table. Records that do not match will be dropped.

2. There were some duplicated codes of creating temporaryTableName and temporaryTableReference. Maybe we can try to combine the previous condition and this one together as much as possible.

Reply:
I think the correct behavior for update operation would be. 
If table does not exists:
   table will be created, but not records will be merged in. (no match)
else if table is empty:
    table will be kept as empty (no match)
else 
     matched records will be updated. drop no matched records.

Reply:
If the Update operation is provided and the table does not exist, it doesn't help to create an empty table as none of the records will match. Also I can not think of any practical usecase where user would want to create an empty table without updating any record in it.
In that case better I think behavior would be to throw an error to the user. 

Though the current behavior is completely wrong. It just dumps every incoming rows to the table. We should create temp table for update operation when table does not exist.

Reply:
There are 3 tests to perform for update and upsert operations:
1.) Table does not exist
2.) Table exists but is empty
3.) Table exists with keys that matches the incoming records. While there are some incoming records that does not match existing keys.

The bug exists for upsert and update operation for test case #1. Rest have correct existing behavior

Reply:
For upsert, in 1. the expected behavior is to create the table, and insert all incoming rows, right?

For update, I'm a little wary of 1. Why should it fail the pipeline? Can't it be a no-op with a warning in the logs?

Reply:
1. Correct. 
2. Had an offline discussion with @bdmogal . We have decided to add a warning message if the table does not exist. The reason is, failing pipeline would not be a good user experience in case many sinks are involved.

Reply:
+1 on not failing the pipeline.

Reply:
Agreed.